src/vm/syscalls/lib.zig

```diff
@@ -1106,6 +1298,285 @@ test createProgramAddress {
+        allocator,
+        prng.random(),
+        .{
+            // Sets total_stake to sum of all voter stakes:
+            .epoch_stakes = &.{
+                .{
+                    .pubkey = target_vote_address,
+                    .stake = total_epoch_stake / 2,
+                },
+                .{
+                    .pubkey = Pubkey.initRandom(prng.random()),
+                    .stake = total_epoch_stake / 2,
+                },
+            },
+        },
+    );
+    defer {
+        ec.deinit();
+        allocator.destroy(ec);
+        sc.deinit();
+        allocator.destroy(sc);
+        tc.deinit();
+    }
+
+    // Test get total stake
+    {
+        tc.compute_meter = tc.compute_budget.syscall_base_cost;
+
+        var memory_map = try MemoryMap.init(allocator, &.{}, .v3, .{});
+        defer memory_map.deinit(allocator);
+
+        var registers = RegisterMap.initFill(0);
+        try getEpochStake(&tc, &memory_map, &registers);
+
+        try std.testing.expectEqual(
+            registers.get(.r0),
+            total_epoch_stake,
+        );
+    }
+
+    // Test invalid read-only memory region.
+    {
+        tc.compute_meter = tc.compute_budget.syscall_base_cost +|
+            (32 / tc.compute_budget.cpi_bytes_per_unit) +|
+            tc.compute_budget.mem_op_base_cost;
+
+        const vote_addr = 0x100000000;
+        const vote_buffer = std.mem.asBytes(&target_vote_address)[0..31]; // not all bytes readable.
+
+        var memory_map = try MemoryMap.init(allocator, &.{
+            memory.Region.init(.constant, vote_buffer, vote_addr),
+        }, .v3, .{});
+        defer memory_map.deinit(allocator);
+
+        var registers = RegisterMap.initFill(0);
+        registers.set(.r1, vote_addr);
+
+        try std.testing.expectError(
+            memory.RegionError.AccessViolation,
+            getEpochStake(&tc, &memory_map, &registers),
+        );
+    }
+
+    // Test valid vote address stake read
+    {
+        tc.compute_meter = tc.compute_budget.syscall_base_cost +|
+            (32 / tc.compute_budget.cpi_bytes_per_unit) +|
+            tc.compute_budget.mem_op_base_cost;
+
+        const vote_addr = 0x100000000;
+        const vote_buffer = std.mem.asBytes(&target_vote_address);
+
+        var memory_map = try MemoryMap.init(allocator, &.{
+            memory.Region.init(.constant, vote_buffer, vote_addr),
+        }, .v3, .{});
+        defer memory_map.deinit(allocator);
+
+        var registers = RegisterMap.initFill(0);
+        registers.set(.r1, vote_addr);
+        try getEpochStake(&tc, &memory_map, &registers);
+
+        try std.testing.expectEqual(
+            registers.get(.r0),
+            total_epoch_stake / 2,
+        );
+    }
+
+    // Test readable address (but not a registered voter).
+    {
+        tc.compute_meter = tc.compute_budget.syscall_base_cost +|
+            (32 / tc.compute_budget.cpi_bytes_per_unit) +|
```

Comment:
I'd prefer this for these just so its clear where the value 32 comes from
```suggestion
            (Pubkey.SIZE / tc.compute_budget.cpi_bytes_per_unit) +|
```

Reply:
https://github.com/Syndica/sig/pull/721/commits/c9290b1874aef839010a2cf304fa87ac4f00b936